packages/strapi-generate-new/lib/merge-template.js

```diff
@@ -0,0 +1,164 @@
+    public: allowChildren,
+    scripts: allowChildren,
+  },
+};
+
+// Traverse template tree to make sure each file and folder is allowed
+async function checkTemplateStructure(templatePath) {
+  // Recursively check if each item in the template is allowed
+  const checkPathContents = (pathToCheck, parents) => {
+    const contents = fse.readdirSync(pathToCheck);
+    contents.forEach(item => {
+      const nextParents = [...parents, item];
+      const matchingTreeValue = _.get(allowedTemplateTree, nextParents);
+
+      // Treat files and directories separately
+      const itemPath = path.resolve(pathToCheck, item);
+      const isDirectory = fse.statSync(itemPath).isDirectory();
+
+      if (matchingTreeValue === undefined) {
+        // Unknown paths are forbidden
+        throw Error(`Illegal template structure, unknown path ${nextParents.join('/')}`);
+      }
+
+      if (matchingTreeValue === true) {
+        if (!isDirectory) {
+          // All good, the file is allowed
+          return;
+        }
+        throw Error(
+          `Illegal template structure, expected a file and got a directory at ${nextParents.join(
+            '/'
+          )}`
+        );
+      }
+
+      if (isDirectory) {
+        if (matchingTreeValue === allowChildren) {
+          // All children are allowed
+          return;
+        }
+        // Check if the contents of the directory are allowed
+        checkPathContents(itemPath, nextParents);
+      } else {
+        throw Error(`Illegal template structure, unknow file ${nextParents.join('/')}`);
+      }
+    });
+  };
+
+  checkPathContents(templatePath, []);
+}
+
+function getRepoInfo(githubURL) {
+  // Make sure it's a github url
+  const address = githubURL.split('://')[1];
+  if (!address.startsWith('github.com')) {
+    throw Error('A Strapi template can only be a GitHub URL');
+  }
+
+  // Parse github address into parts
+  const [, username, name, ...rest] = address.split('/');
+  const isRepo = username != null && name != null;
+  if (!isRepo || rest.length > 0) {
+    throw Error('A template URL must be the root of a GitHub repository');
+  }
+
+  return { username, name };
+}
+
+async function downloadGithubRepo(repoInfo, rootPath) {
+  const { username, name, branch = 'master' } = repoInfo;
+  const codeload = `https://codeload.github.com/${username}/${name}/zip/${branch}`;
+  const templatePath = path.resolve(rootPath, 'tmp-template');
+  const response = await fetch(codeload);
+  await new Promise(resolve => {
+    response.body.pipe(unzip.Extract({ path: templatePath })).on('close', resolve);
+  });
+
+  return templatePath;
+}
+
+// Merge the template's template.json into the Strapi project's package.json
+async function mergePackageJSON(rootPath, templatePath, repoInfo) {
+  // Import the package.json and template.json templates
+  const packageJSON = require(path.resolve(rootPath, 'package.json'));
+  const templateJSON = require(path.resolve(templatePath, 'template.json'));
```

Comment:
the package.json onverwritte should be in a package property inside the template.json file. This will give us more flexibility to extend the template.json config later